x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java

```diff
@@ -210,21 +213,30 @@ public UnresolvedAttribute visitQualifiedName(EsqlBaseParser.QualifiedNameContex
         Source source = source(ctx);
         Literal intLit = typedParsing(this, ctx.integerValue(), Literal.class);
-        Integer value = (Integer) intLit.value();
+        Number value = (Number) intLit.value();
+        if (intLit.dataType() == DataTypes.UNSIGNED_LONG) {
+            value = unsignedLongAsNumber(value.longValue());
+        }
         String qualifier = ctx.UNQUOTED_IDENTIFIER().getText().toLowerCase(Locale.ROOT);
 
-        return switch (qualifier) {
-            case "millisecond", "milliseconds" -> new Literal(source, Duration.ofMillis(value), TIME_DURATION);
-            case "second", "seconds" -> new Literal(source, Duration.ofSeconds(value), TIME_DURATION);
-            case "minute", "minutes" -> new Literal(source, Duration.ofMinutes(value), TIME_DURATION);
-            case "hour", "hours" -> new Literal(source, Duration.ofHours(value), TIME_DURATION);
-
-            case "day", "days" -> new Literal(source, Period.ofDays(value), DATE_PERIOD);
-            case "week", "weeks" -> new Literal(source, Period.ofDays(value * 7), DATE_PERIOD);
-            case "month", "months" -> new Literal(source, Period.ofMonths(value), DATE_PERIOD);
-            case "year", "years" -> new Literal(source, Period.ofYears(value), DATE_PERIOD);
-            default -> throw new ParsingException(source, "Unexpected numeric qualifier '{}'", qualifier);
-        };
+        try {
+            return switch (qualifier) {
+                case "millisecond", "milliseconds" -> new Literal(source, Duration.ofMillis(safeToLong(value)), TIME_DURATION);
+                case "second", "seconds" -> new Literal(source, Duration.ofSeconds(safeToLong(value)), TIME_DURATION);
+                case "minute", "minutes" -> new Literal(source, Duration.ofMinutes(safeToLong(value)), TIME_DURATION);
+                case "hour", "hours" -> new Literal(source, Duration.ofHours(safeToLong(value)), TIME_DURATION);
+
+                case "day", "days" -> new Literal(source, Period.ofDays(safeToInt(safeToLong(value))), DATE_PERIOD);
+                case "week", "weeks" -> new Literal(source, Period.ofWeeks(safeToInt(safeToLong(value))), DATE_PERIOD);
+                case "month", "months" -> new Literal(source, Period.ofMonths(safeToInt(safeToLong(value))), DATE_PERIOD);
+                case "year", "years" -> new Literal(source, Period.ofYears(safeToInt(safeToLong(value))), DATE_PERIOD);
+                default -> throw new ParsingException(source, "Unexpected numeric qualifier '{}'", qualifier);
+            };
+        } catch (QlIllegalArgumentException | ArithmeticException e) {
+            // the range varies by unit: Duration#ofMinutes(), #ofHours() will Math#multiplyExact() to reduce the unit to seconds;
+            // and same for Period#ofWeeks()
+            throw new ParsingException(source, "Number [{}] outside of [{}] range", ctx.integerValue().getText(), qualifier);
+        }
```

Comment:
to avoid repeating `new Literal `, simplify the switch to handle only the value creation.

```
try {
Object value = 
        switch (qualifier) {
            case "millisecond", "milliseconds" -> Duration.ofMillis(numericValue);
            case "second", "seconds" -> Duration.ofSeconds(numericValue);
            case "minute", "minutes" -> Duration.ofMinutes(numericValue);
            case "hour", "hours" -> Duration.ofHours(numericValue);
            case "day", "days" -> Period.ofDays(safeToInt(numericValue));
            case "week", "weeks" -> Period.ofWeeks(safeToInt(numericValue));
            case "month", "months" -> Period.ofMonths(safeToInt(numericValue));
            case "year", "years" -> Period.ofYears(safeToInt(numericValue));
            default -> throw new ParsingException(source, "Unexpected numeric qualifier '{}'", qualifier);
        }
}
} catch (..) {
  throw new Parsing()
}

return new Literal(source(), value, value instanceof Duration ? TIME_DURATION, DATE_PERIOD)
...
```

Reply:
Makes sense, applied.